src/PerfCounters_aarch64.h

```diff
@@ -1,35 +1,180 @@
+      << ", variant: 0x" << int(cpuid.variant) << " part: 0x" << int(cpuid.part);
+  return stm;
+}
 
 /**
  * Return the detected, known microarchitecture of this CPU, or don't
  * return; i.e. never return UnknownCpu.
  */
-static CpuMicroarch compute_cpu_microarch() {
-  FILE *midr_el1 = fopen(midr_path, "r");
-  if (!midr_el1) {
-    CLEAN_FATAL() << "Failed to read midr register from kernel";
-  }
-  uint32_t midir;
-  if (1 != fscanf(midr_el1, "%x", &midir)) {
-    CLEAN_FATAL() << "Failed to read midr register from kernel";
-  }
-  fclose(midr_el1);
-  switch (midir) {
-    case 0x413fd0c1:
+static CpuMicroarch compute_cpu_microarch(const CPUID &cpuid) {
+  switch (cpuid.implementer) {
+  case 0x41: // ARM
+    switch (cpuid.part) {
+    case 0xd0c:
       return ARMNeoverseN1;
-    default:
-      break;
+    }
+    break;
   }
-  CLEAN_FATAL() << "Aarch64 CPU type " << HEX(midir) << " unknown";
+  CLEAN_FATAL() << "Unknown aarch64 CPU type " << cpuid;
   return UnknownCpu; // not reached
 }
 
+static const std::string midr_dir = "/sys/devices/system/cpu/";
+static const std::regex cpuname_regex{"cpu([0-9]+)"};
+
+static inline void get_cpuinfo_sysfs(std::vector<CPUID> &res)
+{
+  auto dir = opendir(midr_dir.c_str());
+  if (!dir)
+    return;
+  while (auto entry = readdir(dir)) {
+    std::cmatch match;
+    if (entry->d_type != DT_DIR ||
+        !std::regex_match(entry->d_name, match, cpuname_regex)) {
+      continue;
+    }
+    auto cpuidx = std::stoul(match[1].str());
+    std::string name = midr_dir + entry->d_name + "/regs/identification/midr_el1";
+    std::ifstream file(name);
+    if (!file) {
+      CLEAN_FATAL() << "Failed to read midr register from kernel";
+    }
+    uint64_t val = 0;
+    file >> std::hex >> val;
+    if (!file) {
+      CLEAN_FATAL() << "Failed to read midr register from kernel";
+    }
+    CPUID cpuid = {
+      uint8_t(val >> 24),
+      uint8_t((val >> 20) & 0xf),
+      uint16_t((val >> 4) & 0xfff)
+    };
+    if (res.size() <= cpuidx) {
+      res.resize(cpuidx + 1);
+    }
+    if (res[cpuidx]) {
+      CLEAN_FATAL() << "Duplicated CPUID for core " << cpuidx;
+    }
+    res[cpuidx] = cpuid;
+  }
+  for (auto &cpuid : res) {
+    if (!cpuid) {
+      CLEAN_FATAL() << "Unable to find CPU id for core " << &cpuid - &res[0];
+    }
+  }
+  closedir(dir);
+}
+
+// Use an external template since lambda's can't be templated in C++11
+template<typename T, typename F>
+static inline bool try_read_procfs_line(const std::string &line,
+                                        const char *prefix, T &out,
+                                        bool &flag, F &&reset)
+{
+  size_t prefix_len = strlen(prefix);
+  if (line.size() < prefix_len)
+    return false;
+  if (memcmp(&line[0], prefix, prefix_len) != 0)
+    return false;
+  if (flag)
+    reset();
+  const char *p = &line[prefix_len];
+  while (*p == '\t' || *p == ' ' || *p == ':')
+    p++;
```

Comment:
Let's do something to prevent us running off the end of the string.

Reply:
It shouldn't happen though? This loop only runs unless it matches one of these character. Hitting a NUL or new line would stop that. (I'll throw in some comments)

Reply:
No I'm saying that I don't see how this would overflow no matter what the buffer contains. Note that I'm not looking for one of the three characters, I'm looking ahead until it's not one of these character. If I run into the end of the string it would also stop.

Reply:
oh, silly me. Sorry!